docs/man/chasquid-util.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "chasquid-util 1"
.TH chasquid-util 1 2025-03-21 "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
chasquid\-util \- chasquid management tool
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBchasquid\-util\fR [\fIoptions\fR] user\-add \fIuser@domain\fR [\-\-password=\fIpassword\fR] [\-\-receive_only]
.PP
\&\fBchasquid\-util\fR [\fIoptions\fR] user\-remove \fIuser@domain\fR
.PP
\&\fBchasquid\-util\fR [\fIoptions\fR] authenticate \fIuser@domain\fR [\-\-password=\fIpassword\fR]
.PP
\&\fBchasquid\-util\fR [\fIoptions\fR] check\-userdb \fIdomain\fR
.PP
\&\fBchasquid\-util\fR [\fIoptions\fR] aliases\-resolve \fIaddr\fR
.PP
\&\fBchasquid\-util\fR [\fIoptions\fR] domaininfo\-remove \fIdomain\fR
.PP
\&\fBchasquid\-util\fR [\fIoptions\fR] print\-config
.SH DESCRIPTION
.IX Header "DESCRIPTION"
chasquid\-util is a command\-line utility for \fBchasquid\fR\|(1) operations.
.SH OPTIONS
.IX Header "OPTIONS"
.IP "\fBuser\-add\fR \fIuser@domain\fR [\-\-password=\fIpassword\fR] [\-\-receive_only]" 8
.IX Item "user-add user@domain [--password=password] [--receive_only]"
Add a new user to the domain.
.Sp
If \fI\-\-receive_only\fR is given, then the user will never successfully
authenticate. This is useful when creating receive\-only users.
.IP "\fBuser\-remove\fR \fIuser@domain\fR" 8
.IX Item "user-remove user@domain"
Remove the user from the domain.
.IP "\fBauthenticate\fR \fIuser@domain\fR [\-\-password=\fIpassword\fR]" 8
.IX Item "authenticate user@domain [--password=password]"
Check the user\*(Aqs password.
.IP "\fBcheck\-userdb\fR \fIdomain\fR" 8
.IX Item "check-userdb domain"
Check the integrity of the domain\*(Aqs users database.
.IP "\fBaliases\-resolve\fR \fIaddr\fR" 8
.IX Item "aliases-resolve addr"
Resolve the given address. Talks to the running chasquid instance.
.IP "\fBdomaininfo\-remove\fR \fIdomain\fR" 8
.IX Item "domaininfo-remove domain"
Remove the domain information entry. This can be used to manually allow a
security level downgrade. Talks to the running chasquid instance.
.IP \fBprint\-config\fR 8
.IX Item "print-config"
Parse and print the configuration in a human\-readable way.
.IP "\fB\-C\fR or \fB\-\-config_dir=<path\fR>" 8
.IX Item "-C or --config_dir=<path>"
Configuration directory.
.IP \fB\-\-configdir=<path\fR> 8
.IX Item "--configdir=<path>"
Deprecated alias for \fB\-\-config_dir=<path\fR>.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBchasquid\fR\|(1)


cmd/chasquid-util/chasquid-util.go
// chasquid-util is a command-line utility for chasquid-related operations.

package main

import (
	"bytes"
	"errors"
	"fmt"
	"io/fs"
	"maps"
	"os"
	"path/filepath"
	"slices"
	"strconv"
	"strings"
	"syscall"

	"blitiri.com.ar/go/chasquid/internal/config"
	"blitiri.com.ar/go/chasquid/internal/envelope"
	"blitiri.com.ar/go/chasquid/internal/localrpc"
	"blitiri.com.ar/go/chasquid/internal/normalize"
	"blitiri.com.ar/go/chasquid/internal/userdb"
	"golang.org/x/term"
	"google.golang.org/protobuf/encoding/prototext"
)

// Usage to show users on --help or invocation errors.
const usage = `
Usage:
  chasquid-util [options] user-add <user@domain> [--password=<password>] [--receive_only]
    Add a user to the userdb.
  chasquid-util [options] user-remove <user@domain>
    Remove a user from the userdb.
  chasquid-util [options] authenticate <user@domain> [--password=<password>]
    Authenticate a user.
  chasquid-util [options] check-userdb <domain>
    Check if the userdb for the given domain is accessible.
  chasquid-util [options] aliases-resolve <address>
    Resolve an address. Talks to the running chasquid.
  chasquid-util [options] domaininfo-remove <domain>
    Remove domaininfo for the given domain. Talks to the running chasquid.
  chasquid-util [options] print-config
    Print the current chasquid configuration.

  chasquid-util [options] dkim-keygen <domain> [<selector> <private-key.pem>] [--algo=rsa3072|rsa4096|ed25519]
    Generate a new DKIM key pair for the domain.
  chasquid-util [options] dkim-dns <domain> [<selector> <private-key.pem>]
    Print the DNS TXT record to use for the domain, selector and
    private key.

Options:
  -C=<path>, --config_dir=<path>  Configuration directory
  -v                              Verbose mode
`

// Command-line arguments.
// Arguments starting with "-" will be parsed as key-value pairs, and
// positional arguments will appear as "$POS" -> value.
//
// For example, "--abc=def x y -p=q -r" will result in:
// {"--abc": "def", "$1": "x", "$2": "y", "-p": "q", "-r": ""}
var args map[string]string

// Globals, loaded from top-level options.
var (
	configDir = "/etc/chasquid"
)

func main() {
	args = parseArgs(usage)

	if _, ok := args["--help"]; ok {
		fmt.Print(usage)
		return
	}

	// Load globals.
	if d, ok := args["--config_dir"]; ok {
		configDir = d
	}
	if d, ok := args["-C"]; ok {
		configDir = d
	}
	if d, ok := args["--configdir"]; ok {
		configDir = d
		Warnf("Option --configdir is deprecated, use --config_dir instead")
	}

	commands := map[string]func(){
		"user-add":          userAdd,
		"user-remove":       userRemove,
		"authenticate":      authenticate,
		"check-userdb":      checkUserDB,
		"aliases-resolve":   aliasesResolve,
		"print-config":      printConfig,
		"domaininfo-remove": domaininfoRemove,
		"dkim-keygen":       dkimKeygen,
		"dkim-dns":          dkimDNS,

		// These exist for testing purposes and may be removed in the future.
		// Do not rely on them.
		"dkim-verify": dkimVerify,
		"dkim-sign":   dkimSign,
	}

	cmd := args["$1"]
	if f, ok := commands[cmd]; ok {
		f()
	} else {
		fmt.Printf("Unknown argument %q\n", cmd)
		Fatalf(usage)
	}
}

// Fatalf prints the given message to stderr, then exits the program with an
// error code.
func Fatalf(s string, arg ...any) {
	fmt.Fprintf(os.Stderr, s+"\n", arg...)
	os.Exit(1)
}

// Warnf prints the given message to stderr, but does not exit the program.
func Warnf(s string, arg ...any) {
	fmt.Fprintf(os.Stderr, s+"\n", arg...)
}

func userDBForDomain(domain string) string {
	if domain == "" {
		domain = args["$2"]
	}
	return configDir + "/domains/" + domain + "/users"
}

func userDBFromArgs(create bool) (string, string, *userdb.DB) {
	username := args["$2"]
	user, domain := envelope.Split(username)
	if domain == "" {
		Fatalf("Domain missing, username should be of the form 'user@domain'")
	}

	if create {
		dbDir := filepath.Dir(userDBForDomain(domain))
		if _, err := os.Stat(dbDir); errors.Is(err, fs.ErrNotExist) {
			fmt.Println("Creating database")
			err = os.MkdirAll(dbDir, 0755)
			if err != nil {
				Fatalf("Error creating database dir: %v", err)
			}
		}
	}

	db, err := userdb.Load(userDBForDomain(domain))
	if err != nil {
		Fatalf("Error loading database: %v", err)
	}

	user, err = normalize.User(user)
	if err != nil {
		Fatalf("Error normalizing user: %v", err)
	}

	return user, domain, db
}

// chasquid-util check-userdb <domain>
func checkUserDB() {
	path := userDBForDomain("")
	// Check if the file exists. This is because userdb.Load does not consider
	// it an error.
	if _, err := os.Stat(path); errors.Is(err, fs.ErrNotExist) {
		Fatalf("Error: file %q does not exist", path)
	}

	udb, err := userdb.Load(path)
	if err != nil {
		Fatalf("Error loading database: %v", err)
	}

	fmt.Printf("Database loaded (%d users)\n", udb.Len())
}

// chasquid-util user-add <user@domain> [--password=<password>] [--receive_only]
func userAdd() {
	user, _, db := userDBFromArgs(true)

	_, recvOnly := args["--receive_only"]
	_, hasPassword := args["--password"]

	if recvOnly && hasPassword {
		Fatalf("Cannot specify both --receive_only and --password")
	}

	var err error
	if recvOnly {
		err = db.AddDeniedUser(user)
	} else {
		password := getPassword()
		err = db.AddUser(user, password)
	}

	if err != nil {
		Fatalf("Error adding user: %v", err)
	}

	err = db.Write()
	if err != nil {
		Fatalf("Error writing database: %v", err)
	}

	fmt.Println("Added user")
}

// chasquid-util authenticate <user@domain> [--password=<password>]
func authenticate() {
	user, _, db := userDBFromArgs(false)

	password := getPassword()
	ok := db.Authenticate(user, password)
	if ok {
		fmt.Println("Authentication succeeded")
	} else {
		Fatalf("Authentication failed")
	}
}

func getPassword() string {
	password, ok := args["--password"]
	if ok {
		return password
	}

	fmt.Printf("Password: ")
	p1, err := term.ReadPassword(syscall.Stdin)
	fmt.Printf("\n")
	if err != nil {
		Fatalf("Error reading password: %v\n", err)
	}

	fmt.Printf("Confirm password: ")
	p2, err := term.ReadPassword(syscall.Stdin)
	fmt.Printf("\n")
	if err != nil {
		Fatalf("Error reading password: %v", err)
	}

	if !bytes.Equal(p1, p2) {
		Fatalf("Passwords don't match")
	}

	return string(p1)
}

// chasquid-util user-remove <user@domain>
func userRemove() {
	user, _, db := userDBFromArgs(false)

	present := db.RemoveUser(user)
	if !present {
		Fatalf("Unknown user")
	}

	err := db.Write()
	if err != nil {
		Fatalf("Error writing database: %v", err)
	}

	fmt.Println("Removed user")
}

// chasquid-util aliases-resolve <address>
func aliasesResolve() {
	conf, err := config.Load(configDir+"/chasquid.conf", "")
	if err != nil {
		Fatalf("Error loading config: %v", err)
	}

	c := localrpc.NewClient(conf.DataDir + "/localrpc-v1")
	vs, err := c.Call("AliasResolve", "Address", args["$2"])
	if err != nil {
		Fatalf("Error resolving: %v", err)
	}

	// Result is a map of type -> []addresses.
	// Sort the types for deterministic output.
	for _, t := range slices.Sorted(maps.Keys(vs)) {
		for _, a := range vs[t] {
			fmt.Printf("%v  %s\n", t, a)
		}
	}
}

// chasquid-util print-config
func printConfig() {
	conf, err := config.Load(configDir+"/chasquid.conf", "")
	if err != nil {
		Fatalf("Error loading config: %v", err)
	}

	fmt.Println(prototext.Format(conf))
}

// chasquid-util domaininfo-remove <domain>
func domaininfoRemove() {
	conf, err := config.Load(configDir+"/chasquid.conf", "")
	if err != nil {
		Fatalf("Error loading config: %v", err)
	}

	c := localrpc.NewClient(conf.DataDir + "/localrpc-v1")
	_, err = c.Call("DomaininfoClear", "Domain", args["$2"])
	if err != nil {
		Fatalf("Error removing domaininfo entry: %v", err)
	}
}

// parseArgs parses the command line arguments, and returns a map.
//
// Arguments starting with "-" will be parsed as key-value pairs, and
// positional arguments will appear as "$POS" -> value.
//
// For example, "--abc=def x y -p=q -r" will result in:
// {"--abc": "def", "$1": "x", "$2": "y", "-p": "q", "-r": ""}
func parseArgs(usage string) map[string]string {
	args := map[string]string{}

	pos := 1
	for _, a := range os.Args[1:] {
		// Note: Consider handling end of args marker "--" explicitly in
		// the future if needed.
		if strings.HasPrefix(a, "-") {
			k, v, _ := strings.Cut(a, "=")
			args[k] = v
		} else {
			args["$"+strconv.Itoa(pos)] = a
			pos++
		}
	}

	return args
}
